### PR TITLE
Add `unlink-image-auto-links` feature

### LINKED ISSUE #4132: Disable image auto-links if they point to themselves
Follow-up of https://github.com/sindresorhus/refined-github/pull/4127#issuecomment-802730145

This would help distinguish useless auto-links from actual, custom links. (The locations of images would still by accessible from the right-click context menu.)

Two examples with auto-linked images (demo screenshots) and purposely-linked images (sponsor logos) :point_down: 
 * https://github.com/sindresorhus/refined-github
 * https://github.com/chalk/chalk

### PR DESCRIPTION
**Pull request by kidonng**
Resolve #4132

## Test URLs

https://github.com/sindresorhus/refined-github
https://github.com/chalk/chalk

### CODE REVIEW

File: source/features/clean-image-auto-links.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+				link.replaceWith(image);
+			}
+		}
+	}
+}
+
+void features.add(__filebasename, {
+	include: [
+		pageDetect.isRepoTree,
+		() => pageDetect.isSingleFile() && location.pathname.endsWith('.md')
```
Comment (yakov116) [inline source/features/clean-image-auto-links.tsx:29]: @fregante add this to detections `isSingleMarkdownFile`? maybe just add to util here?

File: source/features/clean-image-auto-links.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+function init(): void {
+	for (const image of select.all('.markdown-body a > img')) {
+		const link = image.parentElement as HTMLAnchorElement;
+
+		if (image.src === link.href) {
+			link.replaceWith(image);
+		}
+
+		if (pageDetect.isSingleFile(link)) {
+			const rawUrl = new GitHubURL(link.href).assign({route: 'raw'}).toString();
```
Comment (yakov116) [inline source/features/clean-image-auto-links.tsx:16]: does `href` work here?

File: source/features/clean-image-auto-links.tsx (hunk 1 above)
Comment (kidonng) [reply source/features/clean-image-auto-links.tsx:28]: Wait, we already have a better one: https://github.com/sindresorhus/refined-github/blob/3aadf2f9141107d7ca92e8753f2a66cbc10ebd9d/source/features/vertical-front-matter.tsx#L9

File: source/features/clean-image-auto-links.tsx (hunk 2 above)
Comment (kidonng) [reply source/features/clean-image-auto-links.tsx:16]: It works fine? Not sure what are you confirming.

Comment (yakov116) [reply source/features/clean-image-auto-links.tsx:16]: instead of tostring()

File: source/features/clean-image-auto-links.tsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+			if (image.src === rawUrl) {
```
Comment (yakov116) [inline source/features/clean-image-auto-links.tsx:18]:
```suggestion
			const rawUrl = new GitHubURL(link.href).assign({route: 'raw'});

			if (image.src === rawUrl.href) {
```

File: source/features/clean-image-auto-links.tsx (hunk 2 above)
Comment (kidonng) [reply source/features/clean-image-auto-links.tsx:16]: I see, they are identical so it should work.

File: source/features/clean-image-auto-links.tsx [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+function init(): void {
+	for (const image of select.all('.markdown-body a > img')) {
+		const link = image.parentElement as HTMLAnchorElement;
+
+		if (image.src === link.href) {
+			link.replaceWith(image);
+		}
+
+		if (pageDetect.isSingleFile(link)) {
+			const rawUrl = new GitHubURL(link.href).assign({route: 'raw'});
```
Comment (fregante) [inline source/features/clean-image-auto-links.tsx:17]: I’m conflicted about this. The link in this case is not the same and it makes it easier to edit the linked local file

Comment (kidonng) [reply source/features/clean-image-auto-links.tsx:17]: I actually thought this is included in the original requests.

Comment (fregante) [reply source/features/clean-image-auto-links.tsx:17]: Indeed, there are very few situations where this would be useful (e.g. hand-written SVGs)

**Comment by fregante** [comment]
I’d rather merge #4144 in today’s release though 😅

File: readme.md [hunk 5]
```diff
@@ -354,6 +354,7 @@ Thanks for contributing! 🦋🙌
 - [](# "same-page-definition-jump") [Avoids re-loading the page when jumping to function definition in the current file.](https://user-images.githubusercontent.com/16872793/90833649-7a5e2f80-e316-11ea-827d-a4e3ac8ced69.png)
 - [](# "linkify-full-profile-readme-title") [Linkifies the readme text on profile pages.](https://user-images.githubusercontent.com/16872793/90910173-ebe4bf00-e3a4-11ea-8fc5-aea3d1a2e5e5.png)
 - [](# "convert-pr-to-draft-improvements") [Moves the "Convert PR to Draft" button to the mergeability box and adds visual feedback to its confirm button.](https://user-images.githubusercontent.com/1402241/95644892-885f3f80-0a7f-11eb-8428-8e0fb0c8dfa5.gif)
+- [](# "clean-image-auto-links") [Removes unnecessary image auto links if they point to themselves.](https://user-images.githubusercontent.com/44045911/112277464-bb5b5f80-8cbc-11eb-9dc0-0268fc3a2b8a.png)
```
Comment (fregante) [inline readme.md:357]: `unlink-image-auto-links`?

File: source/features/clean-image-auto-links.tsx [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+function init(): void {
+	for (const image of select.all('.markdown-body a > img')) {
+		const link = image.parentElement as HTMLAnchorElement;
+
+		if (image.src === link.href) {
+			link.replaceWith(image);
+		}
+
+		if (pageDetect.isSingleFile(link)) {
```
Comment (fregante) [inline source/features/clean-image-auto-links.tsx:16]: Else if

File: source/features/clean-image-auto-links.tsx (hunk 4 above)
Comment (fregante) [reply source/features/clean-image-auto-links.tsx:17]: If you turn this line into a function, you can avoid having another condition below. Example:

if isSingleFile(link) && image.src === getRawUrl(link)
  link.remove()

File: source/features/clean-image-auto-links.tsx [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+				link.replaceWith(image);
+			}
+		}
+	}
+}
+
+void features.add(__filebasename, {
+	include: [
+		pageDetect.isRepoTree,
+		() => pageDetect.isSingleFile() && isSingleMarkdownFile()
```
Comment (fregante) [inline source/features/clean-image-auto-links.tsx:29]: isSingleMarkdownFile should be enough. It already implies that it’s a single file.

File: source/features/vertical-front-matter.tsx [hunk 8]
```diff
@@ -4,9 +4,7 @@ import select from 'select-dom';
 import * as pageDetect from 'github-url-detection';
 
 import features from '.';
-
-// https://github.com/github/markup/blob/cd01f9ec87c86ce5a7c70188a74ef40fc4669c5b/lib/github/markup/markdown.rb#L34
-const hasFrontMatter = (): boolean => pageDetect.isSingleFile() && /\.(mdx?|mkdn?|mdwn|mdown|markdown|litcoffee)$/.test(location.pathname);
+import {isSingleMarkdownFile} from '../github-helpers';
```
Comment (fregante) [inline source/features/vertical-front-matter.tsx:7]: This isn’t equivalent. Litcoffee isn’t markdown and we need to preserve “has front matter” as its own test

Comment (fregante) [reply source/features/vertical-front-matter.tsx:7]: Given how it’s rendered, I wonder if the test for all of these features should be called “is markdown file” at all: https://github.com/rmetzler/litcoffee-examples/blob/master/fib.litcoffee

There are more file types that would benefit from `unlink-images-auto-link`, like: https://stackoverflow.com/questions/41112069/what-file-types-does-github-support-for-readmes

File: source/features/clean-image-auto-links.tsx (hunk 7 above)
Comment (kidonng) [reply source/features/clean-image-auto-links.tsx]: There are both `isSingleFile` and `isEditingFile` combined with `isSingleMarkdownFile`, so I didn't put `isSingleFile` into `isSingleMarkdownFile`. But perhaps it should be called `isMarkdownFile` from the beginning 😅

File: source/features/unlink-image-auto-links.tsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
+import select from 'select-dom';
+import * as pageDetect from 'github-url-detection';
+
+import features from '.';
+import {isMarkdownFile} from '../github-helpers';
+import GitHubURL from '../github-helpers/github-url';
```
Comment (yakov116) [inline source/features/unlink-image-auto-links.tsx:6]:
```suggestion
import GitHubURL from '../github-helpers/github-url';
import {isMarkdownFile} from '../github-helpers';
```

**Comment by kidonng** [comment]
Maybe this could also apply to profile pages🤔

File: source/features/collapse-markdown-sections.tsx [hunk 10]
```diff
@@ -44,7 +45,7 @@ void features.add(__filebasename, {
 	include: [
 		pageDetect.isRepoTree,
 		pageDetect.isRepoWiki,
-		() => pageDetect.isSingleFile() && location.pathname.endsWith('.md')
+		() => pageDetect.isSingleFile() && isMarkdownFile()
```
Comment (fregante) [inline source/features/collapse-markdown-sections.tsx:48]: It should still not just be limited to markdown files. It's _rendered/markup text files_, almost everywhere in this PR

File: source/features/unlink-image-auto-links.tsx [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+		if (image.src === getRawUrl(link)) {
+			link.replaceWith(image);
+		}
+	}
+}
+
+void features.add(__filebasename, {
+	include: [
+		pageDetect.isRepoTree,
+		() => pageDetect.isSingleFile() && isMarkupFile()
```
Comment (fregante) [inline source/features/unlink-image-auto-links.tsx:43]:
```suggestion
		() => isMarkupFile()
```

`isMarkupFile` should already imply it's single, there should never be a case when `isMarkupFile` is not also a single file

File: source/github-helpers/index.ts [hunk 12]
```diff
@@ -140,3 +140,8 @@ export async function isPermalink(): Promise<boolean> {
 export function isNotRefinedGitHubRepo(): boolean {
 	return !location.pathname.startsWith('/sindresorhus/refined-github/');
 }
+
+// https://github.com/github/markup/blob/cd01f9ec87c86ce5a7c70188a74ef40fc4669c5b/lib/github/markup/markdown.rb#L34
+export function isMarkdownFile(): boolean {
+	return /\.(mdx?|mkdn?|mdwn|mdown|markdown|litcoffee)$/.test(location.pathname);
```
Comment (fregante) [inline source/github-helpers/index.ts:146]: lit coffee is not markdown

**Review by fregante — CHANGES_REQUESTED** [review]

File: source/github-helpers/index.ts (hunk 12 above)
Comment (kidonng) [reply source/github-helpers/index.ts:146]: Quoting [the official website](https://coffeescript.org/#literate):

> If you name your file with a .litcoffee extension, you can write it as a Markdown document

And GitHub's code treats them as Markdown so I see no reason to change.

File: source/features/collapse-markdown-sections.tsx (hunk 10 above)
Comment (kidonng) [reply source/features/collapse-markdown-sections.tsx:48]: I see, so some of these features are actually mistakenly limited to `.md` files when implemented.
Shy we change their names to something else so indicate they work on more files, not just markdown? Or perhaps a separate PR in the future.

File: source/features/unlink-image-auto-links.tsx [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+import GitHubURL from '../github-helpers/github-url';
+import {isMarkdownFile} from '../github-helpers';
+
+function isMarkupFile(): boolean {
+	if (isMarkdownFile()) {
+		return true;
+	}
+
+	// https://github.com/github/markup/blob/46908504f5099b15a88030749f0aeb52cd2f6799/README.md#markups
+	if (/\.(textile|rdoc|org|creole|(media)?wiki|rst|a(scii)?doc|asc|pod)$/.test(location.pathname)) {
```
Comment (fregante) [inline source/features/unlink-image-auto-links.tsx:14]: Let's inline all the related extensions here, without using external "markdown" checks

File: source/github-helpers/index.ts (hunk 12 above)
Comment (fregante) [reply source/github-helpers/index.ts]: Indeed the description for the format is _litera_lly "Literate CS is just Markdown"

**Comment by fregante** [comment]
Can you include more test URLs other than just the root of repositories?

**Comment by fregante** [comment]
Closing this PR for now. Feel free to create a new PR after addressing all the reviews and squashing the commits

**Comment by fregante** [comment]
Ping on this if you have time 😃

**Pull request closed without merging**